File: cpp/137_SingleNumberII.cpp
//with 32 hash
class Solution {
public:
    int singleNumber(vector<int>& nums) {
        vector<int> hash(32,0);
        for(auto num : nums){
            for(int i = 0; i < 32; ++i){
                if((num & (1 << i)) != 0){
                   hash[i] = (hash[i] + 1) % 3; 
                }
            }
        }
        int ret = 0;
        for(int i = 0; i < hash.size(); ++i){
            if(hash[i] > 0){
                ret |= (1 << i);
            }
        }
        return ret;
    }
};

//with 2 variable

File: js/137_SingleNumberII.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var singleNumber = function(nums) {
    var ones = 0, twos = 0;
    for(var i in nums){
        ones = ones ^ nums[i] & ~twos;
        twos = twos ^ nums[i] & ~ones;
    }
    return ones | twos;
};